File: web/selenium/java/src/test/java/com/examples/EvHooks.java
package com.examples;

import static org.junit.Assert.assertFalse;

import com.evinced.EvincedReporter;
import com.evinced.EvincedSDK;
import com.evinced.EvincedWebDriver;
import com.evinced.dto.configuration.EvincedConfiguration;
import com.evinced.dto.results.Report;
import io.github.bonigarcia.wdm.WebDriverManager;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

/**
 * evHooks pattern — per-test lifecycle wrapping with @Before / @After.
 *
 * Each test gets a fresh driver and scan session. The @Before hook starts the
 * continuous scan and @After stops it, saves the report, then tears down the
 * driver — even when the test throws.
 *
 * The Java Selenium SDK does not currently expose a labels API.
 * When a labels API becomes available, add label calls inside @Before to tag
 * each report with the test name, e.g.: evincedDriver.addLabel("test", testName)
 *
 * To upload results to the Evinced Platform, set uploadToPlatform = true in
 * EvincedConfiguration inside setUp().
 */
public class EvHooks {

    private ChromeDriver baseDriver;
    private EvincedWebDriver evincedDriver;

    @Before
    public void setUp() {
        WebDriverManager.chromedriver().setup();
        baseDriver = new ChromeDriver();

        try {
            // Set offline credentials from environment variables
            EvincedSDK.setOfflineCredentials(
                System.getenv("EVINCED_SERVICE_ID"),
                System.getenv("EVINCED_WEB_OFFLINE_TOKEN")
            );

            EvincedConfiguration config = new EvincedConfiguration();
            config.setEnableScreenshots(true);
            // To upload scan results to the Evinced Platform, uncomment the line below:
            // config.setUploadToPlatform(true);

            evincedDriver = new EvincedWebDriver(baseDriver, config);

            // Begin continuous scan for this test
            evincedDriver.evStart();
        } catch (Exception e) {
            baseDriver.quit();
            throw e;
        }
    }

    @After
    public void tearDown() {
        try {
            if (evincedDriver != null) {
                // Stop the scan and retrieve the aggregated report
                Report report = evincedDriver.evStop();

                // Save report using the test method name as the filename prefix
                String timestamp = java.time.LocalDateTime.now()
                    .format(java.time.format.DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss"));
                String baseName = "ev-hooks-report-" + timestamp;

                EvincedReporter.evSaveFile(baseName, report, EvincedReporter.FileFormat.HTML_v2_1);
            }
        } finally {
            if (evincedDriver != null) {
                evincedDriver.quit();
            }
            if (baseDriver != null) {
                baseDriver.quit();
            }
        }
    }

    @Test
    public void landingPageAccessibility() {
        WebDriverWait wait = new WebDriverWait(baseDriver, Duration.ofSeconds(10));

        evincedDriver.get("https://demo.evinced.com/");
        wait.until(ExpectedConditions.presenceOfElementLocated(By.cssSelector("main")));

        // The @After hook will capture and save the report automatically
    }

    @Test
    public void searchFlowAccessibility() {
        WebDriverWait wait = new WebDriverWait(baseDriver, Duration.ofSeconds(10));

        evincedDriver.get("https://demo.evinced.com/");
        wait.until(ExpectedConditions.presenceOfElementLocated(By.cssSelector("main")));

        // Interact with the search form
        evincedDriver.findElement(
            By.xpath("//*[@id='gatsby-focus-wrapper']/main/div[1]/div[3]/div[1]/div/div[1]/p")).click();
        wait.until(ExpectedConditions.presenceOfElementLocated(
            By.xpath("//*[@id='gatsby-focus-wrapper']/main/div[1]/div[3]/div[1]/div/ul/li[3]"))).click();

        evincedDriver.findElement(
            By.xpath("//*[@id='gatsby-focus-wrapper']/main/div[1]/div[3]/div[2]/div/div[1]/p")).click();
        wait.until(ExpectedConditions.presenceOfElementLocated(
            By.xpath("//*[@id='gatsby-focus-wrapper']/main/div[1]/div[3]/div[2]/div/ul/li[3]"))).click();

        evincedDriver.findElement(
            By.xpath("//*[@id='gatsby-focus-wrapper']/main/div[1]/div[3]/a")).click();
        wait.until(ExpectedConditions.presenceOfElementLocated(
            By.xpath("//*[@id='gatsby-focus-wrapper']/main/h1/span")));

        // The @After hook will capture and save the report automatically
    }
}

File: web/selenium/dotnet/EvHooks.cs
using System;
using System.IO;
using NUnit.Framework;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Support.UI;
using Evinced.SDK;

namespace EvincedCSharpTests
{
    /// <summary>
    /// evHooks pattern — per-test lifecycle wrapping with [SetUp] / [TearDown].
    ///
    /// Each test gets a fresh driver and scan session. [SetUp] navigates to the
    /// starting page and begins continuous scanning. [TearDown] stops the scan,
    /// saves the report, and tears down the driver — even when the test throws.
    ///
    /// The .NET Selenium SDK does not currently expose a labels API.
    /// When a labels API becomes available, add label calls inside [SetUp] to tag
    /// each report with the test name.
    ///
    /// To upload results to the Evinced Platform, set EvincedSDK.UploadToPlatform = true
    /// inside EvincedConfig.Initialize() or directly in [SetUp].
    /// </summary>
    [TestFixture]
    public class EvHooksTest
    {
        private ChromeDriver baseDriver = null!;
        private IEvincedDriver driver = null!;

        [SetUp]
        public void SetUp()
        {
            // Reads EVINCED_SERVICE_ID and EVINCED_WEB_OFFLINE_TOKEN from environment variables
            EvincedConfig.Initialize();

            baseDriver = new ChromeDriver();
            driver = EvincedDriverFactory.Create(baseDriver);

            // Navigate to the starting page before beginning the scan so the SDK
            // can inject into an already-loaded document.
            WebDriverWait wait = new WebDriverWait(baseDriver, TimeSpan.FromSeconds(10));
            driver.Navigate().GoToUrl("https://demo.evinced.com/");
            wait.Until(d => d.FindElement(By.CssSelector("main")));

            // Begin continuous scan for this test
            driver.EvStart();
        }

        [TearDown]
        public void TearDown()
        {
            try
            {
                if (driver != null)
                {
                    // Stop the scan and retrieve the aggregated report
                    IReport report = driver.EvStop();

                    // Save HTML and JSON reports to disk
                    string timestamp = DateTime.Now.ToString("yyyyMMdd-HHmmss");
                    string baseName = $"ev-hooks-report-{timestamp}";
                    EvincedSDK.EvSaveFile(baseName, report, FileFormat.JSON);
                    EvincedSDK.EvSaveFile(baseName, report, FileFormat.HTML, driver);

                    Console.WriteLine("Reports saved to: " + Directory.GetCurrentDirectory());
                }
            }
            finally
            {
                // Use baseDriver.Quit() only — driver.Quit() (the proxy) calls
                // stopAnalysis() internally and would double-stop after EvStop().
                baseDriver?.Quit();
            }
        }

        [Test]
        public void LandingPageAccessibility()
        {
            // Page was already loaded in SetUp. Nothing else to do here —
            // [TearDown] stops the scan and saves the report automatically.
        }

        [Test]
        public void SearchFlowAccessibility()
        {
            WebDriverWait wait = new WebDriverWait(baseDriver, TimeSpan.FromSeconds(10));

            // Open the "What type" dropdown and select an option.
            // Only one dropdown is filled so the demo site does not auto-navigate.
            var whatDropdown = wait.Until(d => d.FindElement(
                By.XPath("//*[@id='gatsby-focus-wrapper']/main/div[1]/div[3]/div[1]/div/div[1]/p")));
            whatDropdown.Click();

            var treehouseOption = wait.Until(d => d.FindElement(
                By.XPath("//*[@id='gatsby-focus-wrapper']/main/div[1]/div[3]/div[1]/div/ul/li[3]")));
            treehouseOption.Click();

            var whereDropdown = wait.Until(d => d.FindElement(
                By.XPath("//*[@id='gatsby-focus-wrapper']/main/div[1]/div[3]/div[2]/div/div[1]/p")));
            whereDropdown.Click();

            // [TearDown] will stop the scan and save the report automatically
        }
    }
}
